**1. Создание и изменение массивов**

In [2]:
import numpy as np
arr = np.arange(1, 10).reshape(3, 3)
arr *= 2
arr[arr > 10] = 0

print(arr)

[[ 2  4  6]
 [ 8 10  0]
 [ 0  0  0]]


**2. Работа с булевыми масками**

In [5]:
arr = np.random.randint(1, 101, size=20)

mask = arr % 5 == 0
div_by_5 = arr[mask]

arr[mask] = -1

arr, div_by_5

(array([11, -1, 11, 18, -1, 66, 21, -1, 37, 33, -1, 37, 84, -1, 68, -1, 46,
        91, 32,  9]),
 array([10, 85, 65,  5, 65, 45]))

**3. Объединение и разбиение массивов**

In [18]:
arr1 = np.random.randint(0, 51, size=(1, 5))
arr2 = np.random.randint(0, 51, size=(1, 5))

print(arr1, arr2, "\n")

com_arr = np.vstack((arr1, arr2))

print(com_arr, "\n")

split_arr1, split_arr2 = np.vsplit(com_arr, 2)
print("Разделенный массив 1:", split_arr1)
print("Разделенный массив 2:", split_arr2)

[[21 21 22 33 35]] [[15  1 26 16 27]] 

[[21 21 22 33 35]
 [15  1 26 16 27]] 

Разделенный массив 1: [[21 21 22 33 35]]
Разделенный массив 2: [[15  1 26 16 27]]


**4. Генерация и работа с линейными последовательностями**

In [24]:
arr = np.linspace(-10, 10, 50)

sum_el = np.sum(arr)
sum_positive = np.sum(arr[arr > 0])
sum_negative = np.sum(arr[arr < 0])

print("Массив:", arr)
print("Сумма всех элементов:", sum_el)
print("Сумма положительных элементов:", sum_positive)
print("Сумма отрицательных элементов:", sum_negative)

Массив: [-10.          -9.59183673  -9.18367347  -8.7755102   -8.36734694
  -7.95918367  -7.55102041  -7.14285714  -6.73469388  -6.32653061
  -5.91836735  -5.51020408  -5.10204082  -4.69387755  -4.28571429
  -3.87755102  -3.46938776  -3.06122449  -2.65306122  -2.24489796
  -1.83673469  -1.42857143  -1.02040816  -0.6122449   -0.20408163
   0.20408163   0.6122449    1.02040816   1.42857143   1.83673469
   2.24489796   2.65306122   3.06122449   3.46938776   3.87755102
   4.28571429   4.69387755   5.10204082   5.51020408   5.91836735
   6.32653061   6.73469388   7.14285714   7.55102041   7.95918367
   8.36734694   8.7755102    9.18367347   9.59183673  10.        ]
Сумма всех элементов: 7.105427357601002e-15
Сумма положительных элементов: 127.55102040816328
Сумма отрицательных элементов: -127.55102040816327


**5. Работа  с диагональными и единичными матрицами**

In [45]:
arr_eye = np.eye(4)
arr_eye_dig = np.diag([5, 10, 15, 20])

sum_arr_eye = np.sum(arr_eye)
sum_arr_eye_dig = np.sum(arr_eye_dig)

print("Единичная матрица:\n", arr_eye)
print("Сумма элементов единичной матрицы:", sum_arr_eye)

print("\nДиагональная матрица:\n", arr_eye_dig)
print("Сумма элементов диагональной матрицы:", sum_arr_eye_dig)

Единичная матрица:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
Сумма элементов единичной матрицы: 4.0

Диагональная матрица:
 [[ 5  0  0  0]
 [ 0 10  0  0]
 [ 0  0 15  0]
 [ 0  0  0 20]]
Сумма элементов диагональной матрицы: 50


**6. Создание и базовые операции с матрицами**

In [54]:
matrix1 = np.matrix(np.random.randint(1, 21, (3, 3)))
matrix2 = np.matrix(np.random.randint(1, 21, (3, 3)))

sum_matrix = matrix1 + matrix2
diff_matrix = matrix1 - matrix2
prod_matrix = np.multiply(matrix1, matrix2) 

print("Первая матрица:\n", matrix1)
print("Вторая матрица:\n", matrix2)
print("Сумма матриц:\n", sum_matrix)
print("Разность матриц:\n", diff_matrix)
print("Поэлементное произведение матриц:\n", prod_matrix)


Первая матрица:
 [[ 6  8 15]
 [ 4  9 19]
 [ 2  9 14]]
Вторая матрица:
 [[ 9  9 10]
 [19  1 11]
 [ 4  9  1]]
Сумма матриц:
 [[15 17 25]
 [23 10 30]
 [ 6 18 15]]
Разность матриц:
 [[ -3  -1   5]
 [-15   8   8]
 [ -2   0  13]]
Поэлементное произведение матриц:
 [[ 54  72 150]
 [ 76   9 209]
 [  8  81  14]]


**7. Умножение матриц**

In [59]:
matrix1 = np.matrix(np.random.randint(1, 11, (2, 3)))
matrix2 = np.matrix(np.random.randint(1, 11, (3, 2)))

dot_matrix = matrix1 @ matrix2

print("Первая матрица:\n", matrix1)
print("Вторая матрица:\n", matrix2)
print("Результат матричного умножения:\n", dot_matrix)

Первая матрица:
 [[4 7 2]
 [4 6 7]]
Вторая матрица:
 [[10  1]
 [ 7  6]
 [ 2  2]]
Результат матричного умножения:
 [[93 50]
 [96 54]]


**8. Определитесь и обратная матрица**

In [104]:
matrix = np.matrix(np.random.randint(1, 11, (3, 3)))
determ = np.linalg.det(matrix)

if np.isclose(determ, 0):
    print("Матрица вырожденна")
else:
    print("Обратная матрица:\n", np.linalg.inv(matrix))

Обратная матрица:
 [[-0.61111111  0.62962963 -0.09259259]
 [ 0.88888889 -0.7037037   0.07407407]
 [ 0.22222222 -0.25925926  0.18518519]]


**9. Транспонирование и след матрицы**

In [108]:
matrix = np.matrix(np.random.randint(1, 51, (4, 4)))

transpons_matrix = matrix.T

trace_matrix = np.trace(matrix)

print("Исходная матрица:\n", matrix)
print("Транспонированная матрица:\n", transpons_matrix)
print("След матрицы:", trace_matrix)

Исходная матрица:
 [[49 47 31 18]
 [42 21 34 50]
 [44  4  3 38]
 [43 25  7  6]]
Транспонированная матрица:
 [[49 42 44 43]
 [47 21  4 25]
 [31 34  3  7]
 [18 50 38  6]]
След матрицы: 79


**10. Системы линейных уравнений**

In [114]:
A = np.array([
    [2, 3, -1],
    [4, -1, 2],
    [-3, 5, 4]
])
B = np.array([5, 6, -2])

x = np.linalg.solve(A, B)

print("Решение системы:\n", x)

Решение системы:
 [1.63963964 0.57657658 0.00900901]


**11. Индивидуальное задание**

8. Перераспределение бюджета. Городской бюджет распределяется между
образованием, здравоохранением и транспортом. На здравоохранение
выделяется в 1.5 раза больше, чем на транспорт, а на образование — на 20%
больше, чем на здравоохранение. Всего в бюджете 5 000 000 рублей. Сколько
денег выделено на каждую сферу?

X - транспорт; Y - здравоохранение; Z - образование


y =  1.5x

z = 1.2y

x + y + z = 5000000

*Система уравнений:*

\begin{cases} 
x + y + z = 5 000 000 \\
-1.5x + y = 0 \\ 
-1.2y + z = 0 
\end{cases}


In [260]:
A = np.array([
    [1, 1, 1,],
    [-1.5, 1, 0],
    [0, -1.2, 1]
])

B = np.array([5000000, 0, 0])

det_A = np.linalg.det(A)

A_x = A.copy()
A_x[:, 0] = B
det_A_x = np.linalg.det(A_x)

A_y = A.copy()
A_y[:, 1] = B
det_A_y = np.linalg.det(A_y)

A_z = A.copy()
A_z[:, 2] = B
det_A_z = np.linalg.det(A_z)

x = det_A_x / det_A
y = det_A_y / det_A
z = det_A_z / det_A

print("Метод крамера:\n")
print(f"x = {x:.2f} рублей")
print(f"y = {y:.2f} рублей")
print(f"z = {z:.2f} рублей")

Метод крамера:

x = 1162790.70 рублей
y = 1744186.05 рублей
z = 2093023.26 рублей


In [262]:
results = np.linalg.inv(A) @ B

print("Решение матричным методом:\n")
print(f"x: {results[0]:.2f} рублей")
print(f"y: {results[1]:.2f} рублей")
print(f"z: {results[2]:.2f} рублей")

Решение матричным методом:

x: 1162790.70 рублей
y: 1744186.05 рублей
z: 2093023.26 рублей


In [276]:
results_solve = np.linalg.solve(A, B)

print("Решение с помощью np.linalg.solve:\n")
print(f"x: {results_solve[0]:.2f} рублей")
print(f"y: {results_solve[1]:.2f} рублей")
print(f"z: {results_solve[2]:.2f} рублей")

Решение с помощью np.linalg.solve:

x: 1162790.70 рублей
y: 1744186.05 рублей
z: 2093023.26 рублей


**Все методы дали одинаковый результат, т.е. ответ:**

Транспорт: 1162790.70 руб.

Здравоохранение: 1744186.05 руб.

Образование: 2093023.26 руб.